app/controllers/application_controller/buttons.rb

```diff
@@ -147,7 +177,58 @@ def automate_button_field_changed
         page.replace("form_role_visibility", :partial => "layouts/role_visibility", :locals => {:rec_id => (@custom_button.id || "new").to_s, :action => "automate_button_field_changed"})
       end
       unless params[:target_class]
-        @changed = session[:changed] = (@edit[:new] != @edit[:current])
+        @changed = session[:changed] = false
+
+        # This block checks if any of the fields have changed
+        # Broken into different branches for specific fields
+        # If no fields are changed it will not set @changed or session[:changed] to true
+        @edit[:new].each_key do |key|
+          if @edit[:new][key] != @edit[:current][key]
+            if @edit[:new][key] == "" && @edit[:current][key].nil? # check empty string / nil case
+              next
+            elsif @edit[:new][key] == @edit[:current][key].to_s # check string / integer case
+              next
+            elsif key == :roles # check role values
+              if (@edit[:new][key] == ["_ALL_"] || @edit[:new][key] == []) && (@edit[:current][key].nil? || @edit[:current][key] == ["all"] || @edit[:current][key] == ["_ALL_"]) # if visibility is set to "To All"
+                next
+              elsif @edit[:new][key].sort == @edit[:current][key].sort # check if new roles array and current roles array are equal after sorting the ids
```

Comment:
I vaguely recall that the ids could be numbers as strings... do we need to do something like:

```ruby
@edit[:new][key].collect(&:to_i).sort == @edit[:current][key].collect(&:to_i).sort
```

Preferably we fix the code that sets up these hashes so it's always numerics but if there can be strings in there, maybe we can hack it for now.